[11.x] Rehash user passwords when validating credentials

valorin opened the pull request:
## Summary

As was discovered after updating bcrypt rounds from 10 to 12 in https://github.com/laravel/laravel/pull/6245 and https://github.com/laravel/framework/pull/48494, user passwords are not currently being rehashed during login when the hashing configuration has changed. This should be considered a security risk as rehashing passwords is an important part of the authentication process. Rehashing passwords during login ensures updates to the hashing configuration, such as increasing rounds/cost, is applied to existing hashes when the plaintext password is available during login.

This PR implements rehashing directly within the Eloquent and Database User Providers, making it a part of the core framework authentication system. The rehashing will happen automatically when user credentials are validated by `*UserProvider::validateCredentials()`, which is called by the `Auth::attempt()` helper. This helper is used by both Breeze and Fortify/Jetstream to authenticate users, and is the recommended method in the Laravel docs. This ensures that with Laravel 11, any apps using the authentication system will be properly rehashing passwords.

I modelled the new `DatabaseUserProvider::rehashUserPassword()` method on the existing `rehashUserPassword()` method, to ensure it is compatible for apps that don't use Eloquent.

## Breaking Changes

Since rehashing needs to update the database, it needs to know the name of the password attribute on the user model. There is not an existing way to retrieve this information, so I had to add the `getAuthPasswordName()` method to the `Authenticatable` contract. I've implemented it within the `Authenticatable` trait and `GenericUser` so most apps won't be impacted, but apps that don't _use_ or _extend_ these two, or have a different password attribute, will need implement this method.

This is the reason I'm targeting Laravel 11. I'd **love** to get it into 10, but I can't see a way without breaking some obscure configuration. It's not a critical security risk, so it doesn't require an immediate fix.

## Upgrade Impacts

1. Rehashing passwords incurs a performance hit, and after the upgrade to 11, every user who logs in will trigger a password rehash. This is unlike to have an impact as logins are generally spread out and servers usually have spare resources. However apps with large numbers of concurrent logins and limited resources may notice increased resource usage.
1. It's not technically a breaking change, but when a password is rehashed it will force all other active sessions, remember-me tokens, and anything else that relies on the current password hash to be reset automatically. This may cause an increase in login attempts as users reauthenticate different devices.

## Patching Laravel 10 

Since I know it'll be a question asked many times, this change cannot be directly patched in Laravel 10 due to the BC break. 

If you use Laravel Breeze, you can easily patch it to perform rehashing as it copies auth code to the user's app. This isn't directly possible within Jetstream/Fortify, as that code is within the package. There is no point patching these now when a better fix is coming in v11.

<details>
  <summary>Patching Breeze Instructions</summary>
  
1. Open the `\App\Http\Requests\Auth\LoginRequest` file
2. Find the `authenticate()` method
3. Find the `RateLimiter::clear($this->throttleKey());` and add the following:

```
$user = Auth::user();

if (Hash::needsRehash($user->password)) {
    $user->forceFill(['password' => Hash::make($this->input('password'))])->save();
}
```

The method should look something like this:

```
public function authenticate(): void
{
    $this->ensureIsNotRateLimited();

    if (! Auth::attempt($this->only('email', 'password'), $this->boolean('remember'))) {
        RateLimiter::hit($this->throttleKey());

        throw ValidationException::withMessages([
            'email' => trans('auth.failed'),
        ]);
    }

    RateLimiter::clear($this->throttleKey());

    $user = Auth::user();

    if (Hash::needsRehash($user->password)) {
        $user->forceFill(['password' => Hash::make($this->input('password'))])->save();
    }
}
```

</details>

valorin committed:
Rehash user passwords when validating credentials

valorin committed:
Fix style violations

PerryvanderMeer commented:
Does this approach update `update_at` with every successful login attempt? That may not be the desired behavior?

valorin commented:
It'll only update when the password is rehashed, which will happen once.

chrysanthos commented:
@valorin can we potentially use `method_exists` to check if `getAuthPasswordName` method exists on a model and do the rehashing only then? This way we would make this an opt-in feature for L10 and add it in the contract for L11.

valorin commented:
Oooh, that's a great idea!

I'd like to check with the Laravel team to make sure it's a pattern they want, but I'd love to do it if we can. 😁

claudiodekker commented:
Personally I think you can just use the 'password' string instead of a separate method to resolve the field name, but that's me.

While it's more flexible this way, I just can't foresee anyone using anything different, especially since Laravel internally already assumes the 'password' field to exist on the default `Authenticatable` implementation:

https://github.com/laravel/framework/blob/cddb4f3bb5231f44f18fce304dbf190ad8348ddc/src/Illuminate/Auth/Authenticatable.php#L43-L53

claudiodekker commented:
> @valorin can we potentially use `method_exists` to check if `getAuthPasswordName` method exists on a model and do the rehashing only then? This way we would make this an opt-in feature for L10 and add it in the contract for L11.

Yeah, that's probably a better approach as to not break implementations that rely on the current version of the `Authenticatable` contract, since they wouldn't have that new `getAuthPasswordName` method. 👍 

Besides, I doubt anyone has a reason to change the field name from `password` pretty much ever anyway? It's not really a heavily-opinionated name for a field IMO.

valorin committed:
Remove hardcoded password when it's changable

valorin commented:
Converting this to draft as it depends on the status of https://github.com/laravel/framework/pull/48673 and changes will be required if that's merged.

valorin commented:
Converting this to draft as it depends on the status of https://github.com/laravel/framework/pull/48673, which applies the rehashing to Laravel 10 in a backwards compatible way. If that PR is merged, then changes will be required to clean up and finish the upgrade.

roycewilliams commented:
Naive question - where in the code are the criteria for checking whether an upgrade is needed? Is it keyed on any hash with bcrypt cost less than the current standard, or something else?

valorin commented:
@roycewilliams 

It first runs this to check if it needs to rehash:

```
if ($this->hasher->needsRehash($hash)) {
    $this->rehashUserPassword($user, $plain);
}
```

The rehash is triggered if the hashing configuration is changed, which is defined in the application configuration and needs to be manually changed in the app to be triggered. Any changes to that will trigger a rehash though, even if you go down. but it's not something you change randomly. So only apps that specifically change it will have rehashing triggered.

crynobone commented:
I believe this may also create breaking change for any 3rd party authentication package such as https://github.com/DirectoryTree/LdapRecord-Laravel/blob/master/src/Auth/DatabaseUserProvider.php 

I do believe if we want to introduce this in a minor release it should be opt-in to avoid breaking existing projects.

deleugpn commented:
> Besides, I doubt anyone has a reason to change the field name from password pretty much ever anyway?

Reason 1: A company starts a PHP project in 2006 before Composer even existed. There are millions of PHP code without automation tests. In 2017 the company starts using Laravel side-by-side with the original application. It's not a brand new database. You have to comply with a database that was modelled in 2006.

Reason 2: A government contract with local regulations requires the "storage layer which will hold citizen information MUST be defined using the country's local language"

Reason 3: A 2019 Laravel project that migrated from Bcrypt to Argon2id decided to create a new column, rehash upon login and keep both hashes in parallel during rollout until they were confident in dropoping the bcrypt hash, but now the project just runs with a `password_v2` column 🤷 

I could probably remember of another reason, but I think I've made my point 😅

bigship-prashant commented:
I thing we need to give config key in auth.php for developers to enable rehashing or not

trevorgehman commented:
Just adding my two cents: I think it's a no-brainer that this should be in the default Laravel authentication scaffolding. It's a very simple way to ensure security best-practices are being followed, which greatly benefits the Laravel ecosystem as a whole.

driesvints commented:
@valorin guess we can move this out of draft now?

valorin commented:
I've got some changes to make before it can go out of draft. Been a busy few weeks, but I'll try and get them done this week. 🤞

valorin committed:
Shift rehashing into SessionGuard

The Session guard's attempt() method is a better place to apply
rehashing than the validateCredentials() method on the provider.
The latter shouldn't have side-effects, as per it's name.

valorin committed:
Fix style violation

valorin committed:
Add config option to disable rehashing on login

valorin committed:
Clean up rehash flag injection

valorin committed:
Fix contract in DatabaseUserProvider

valorin commented:
This is good to review and (hopefully) merge into 11.x. 🤞
I've made the following changes:

1. I moved the code that triggers the rehashing into `SessionGuard::attempt()` and out of `validateCredentials()` on the user provider. This feels like a better place for it, as rehashing is part of the login attempt, rather than the credential validation itself.
2. I've added a config key `hashing.rehash_on_login` which toggles the rehashing, so it can be disabled if needed. This may be useful for custom auth scaffolding. I've PR'ed the config key: https://github.com/laravel/laravel/pull/6282

I've tested this change with both Breeze and Jetstream and they both successfully rehashed the password on login.

Something interesting I noticed is `SessionGuard` does include a different `rehashUserPassword()` method: https://github.com/laravel/framework/blob/master/src/Illuminate/Auth/SessionGuard.php#L682-L700
However this appears to only be used for Jetstream (as part of logging out other devices), and assumes the EloquentUserProvider is being used, so it wouldn't be useful in this case. My implementation is supposed to work across all User Providers and password attributes.

Shkeats commented on src/Illuminate/Auth/EloquentUserProvider.php:
```diff
@@ -155,6 +155,24 @@ public function validateCredentials(UserContract $user, array $credentials)
+     *
+     * @param  \Illuminate\Contracts\Auth\Authenticatable  $user
+     * @param  array  $credentials
+     * @return string|null
```
need to recheck the return type docblocks on the three concrete versions of this method, the contract one is correct.

valorin replied:
Nice catch! Fixed. 👍

taylorotwell commented:
@valorin what is the behavior of this when the `AuthenticateSession` middleware is enabled? I assume the user is still able to stay logged in if they login and their password is immediately rehashed? Are other sessions logged out?

valorin committed:
Fixing return type in the docblocks

valorin commented:
@taylorotwell Yep, just confirmed with Breeze. The password is rehashed on login (once) and the user is successfully logged in, while other sessions are ended. 

There is no way to avoid the other sessions being ended, but it makes sense that it's triggered by a login rather than just randomly happening, and the user should hopefully associate the two events and not be concerned. It'll also only happen once.

valorin committed:
Use hash_equals() for a secure string comparison

taylorotwell committed:
formatting

taylorotwell committed:
formatting, leverage method on logoutOtherDevices

chrysanthos commented on src/Illuminate/Auth/SessionGuard.php:
```diff
@@ -97,7 +97,7 @@ class SessionGuard implements StatefulGuard, SupportsBasicAuth
 
     /**
-     * Rehash passwords during login.
+     * Indicates if paswords should be rehashed on login if needed.
```
```suggestion
     * Indicates if passwords should be rehashed on login if needed.
```

valorin replied:
Thanks!

valorin committed:
Fix spelling of passwords

Co-authored-by: Chrysanthos <48060191+chrysanthos@users.noreply.github.com>

akulmehta commented:
@valorin would this also work with apps running [Laravel UI](https://github.com/laravel/ui)?

valorin commented:
Yep, it should work after upgrading to 11. 👍 

From what I can see in the code, UI uses `Auth::guard()->attempt()`, which  calls `\Illuminate\Auth\SessionGuard::attempt()` (when you're using a session guard), and that includes the rehashing code.
